server/controller_client.go

```diff
@@ -0,0 +1,231 @@
+// Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+// or more contributor license agreements. Licensed under the Elastic License;
+// you may not use this file except in compliance with the Elastic License.
+
+package server
+
+import (
+	"context"
+	"fmt"
+	"sync"
+
+	"github.com/elastic/elastic-agent-client/v7/pkg/client"
+	"github.com/elastic/elastic-agent-libs/logp"
+	"github.com/elastic/elastic-agent-shipper/config"
+)
+
+type doneChan chan struct{}
+
+// A little helper for managing the state of the main runloop
+type clientHandler struct {
+	// tells the shipper server to begin shutdown
+	shutdownInit doneChan
+	// Tells the controller that the shipper backend has gracefully shut down.
+	shutdownComplete sync.WaitGroup
+
+	unitsMut sync.Mutex
+	units    map[string]*client.Unit
+
+	// the unit ID for the shipper itself
+	shipperOutputID string
+
+	shipperIsStopping bool
+
+	log *logp.Logger
+}
+
+func newClientHandler() clientHandler {
+	return clientHandler{
+		shutdownInit: make(doneChan, 1),
+		log:          logp.L(),
+		units:        make(map[string]*client.Unit),
+	}
+}
+
+func (c *clientHandler) addUnit(unit *client.Unit) {
+	c.unitsMut.Lock()
+	c.units[unit.ID()] = unit
+	c.unitsMut.Unlock()
+}
+
+func (c *clientHandler) getUnit(ID string) *client.Unit {
+	c.unitsMut.Lock()
+	defer c.unitsMut.Unlock()
+	return c.units[ID]
+
+}
+
+func (c *clientHandler) deleteUnit(unit *client.Unit) {
+	c.unitsMut.Lock()
+	delete(c.units, unit.ID())
+	c.unitsMut.Unlock()
+}
+
+// We'll need to track the client ID that's used for the shipper backend itself.
+func (c *clientHandler) setShipperUnitID(unit *client.Unit) {
+	c.shipperOutputID = unit.ID()
+}
+
+// async stop of the shipper's GRPC server, and anything else it needs to manually shutdown
+func (c *clientHandler) stopShipper() {
+	c.log.Debugf("Stopping Shipper")
+	c.shipperIsStopping = true
```

Comment:
This does not look thread-safe, the `startShipper` function can be in race condition with `shutdown` and `stopShipper` because access to `shipperIsStopping` is not synchronised. Or it's not possible to have them run at the same time?

We could switch to uint32 for `c.shipperIsStopping` and using `atomic.CompareAndSwapUint32`.

Reply:
So, my understanding is that we won't get a shutdown command until we're in a `healthy` state, and vice-versa with startup, but I agree that bool isn't the best here, and there's enough unknowns in how the server will behave that we should probably just make sure they're synchronized.